### PR TITLE
CMR-5659

### PR DESCRIPTION
**Pull request by ygliuvt**
As an operator, I would like to refresh the fingerprint value of variable concepts.

### CODE REVIEW

File: bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj [hunk 1]
```diff
@@ -127,6 +129,16 @@
   [context dispatcher provider-id entry-title]
   (dispatch/bootstrap-virtual-products dispatcher context provider-id entry-title))
 
+(defn fingerprint-by-id
+  "Update the fingerprint of the given variable if necessary."
+  [context dispatcher concept-id]
+   (dispatch/fingerprint-by-id dispatcher context concept-id))
```
Comment (chris-durbin) [inline bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj]: formatting off

File: bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj [hunk 2]
```diff
@@ -127,6 +129,16 @@
...
+
+(defn fingerprint-variables
+  "Update the fingerprint of the variables specified by the given params if necessary."
+  [context dispatcher params]
+   (dispatch/fingerprint-variables dispatcher context params))
```
Comment (chris-durbin) [inline bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj]: formatting off

File: system-int-test/src/cmr/system_int_test/utils/url_helper.clj [hunk 3]
```diff
@@ -428,6 +428,20 @@
...
 
+(defn fingerprint-url
+  ([]
+   (format "http://localhost:%s/fingerprint/variables" (transmit-config/bootstrap-port)))
+  ([concept-id]
+   (format "%s/%s" (fingerprint-url) concept-id)))
+
+(defn fingerprint-by-provider-url
+  [provider-id]
+  (format "%s?provider=%s&synchronous=true" (fingerprint-url) provider-id))
```
Comment (chris-durbin) [inline system-int-test/src/cmr/system_int_test/utils/url_helper.clj]: These are meant to be URLs with no query parameters. We should be passing in query parameters when we make a request.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj [hunk 4]
```diff
@@ -94,6 +95,31 @@
          vals
          first))))
 
+(defn- append-stmt
```
Comment (chris-durbin) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: Can we do a GTM to review these database query changes? I'm wondering why these changes are needed, and not sure about doing string manipulation on SQL.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj [hunk 5]
```diff
@@ -94,6 +95,31 @@
...
+      (format "%s and %s" stmt-str part)
+      (format "%s where %s" stmt-str part))))
+
+(defn- find-batch-starting-id-with-stmt
+  "Returns the first id that would be returned in a batch."
+  ([conn stmt]
+   (find-batch-starting-id-with-stmt conn stmt 0))
+  ([conn stmt min-id]
+   (let [stmt (-> stmt
+                  ; string/lower-case
```
Comment (chris-durbin) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: Remove the commented out code (if we keep these functions).

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj [hunk 6]
```diff
@@ -192,6 +218,44 @@
        (when-let [start-index (find-batch-starting-id db table params)]
          (lazy-find (max requested-start-index start-index)))))))
 
+(defn find-concepts-in-batches-with-stmt
```
Comment (chris-durbin) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: I recognize this code from elsewhere, seems like a copy/paste with a tiny bit of change. Let's see if there's an alternative.

File: system-int-test/src/cmr/system_int_test/utils/bootstrap_util.clj [hunk 7]
```diff
@@ -162,6 +162,41 @@
          body (json/decode (:body response) true)]
      (assoc body :status (:status response)))))
 
+(defn- fingerprint-by-url
```
Comment (chris-durbin) [inline system-int-test/src/cmr/system_int_test/utils/bootstrap_util.clj]: Here's the place where you would add the query parameters that you want to use in the tests like synchronous=true.

File: system-int-test/test/cmr/system_int_test/bootstrap/fingerprint_test.clj [hunk 8]
```diff
@@ -0,0 +1,229 @@
+(ns cmr.system-int-test.bootstrap.fingerprint-test
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/fingerprint_test.clj]: Really good tests.

File: bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj [hunk 9]
```diff
@@ -26,6 +26,8 @@
    :index-collection :core-async-dispatcher
    :index-system-concepts :core-async-dispatcher
    :index-concepts-by-id :core-async-dispatcher
+   :fingerprint-by-id :synchronous-dispatcher
+   :fingerprint-variables :core-async-dispatcher
```
Comment (chris-durbin) [inline bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj]: Using a core-async dispatcher means we can't distribute the work across multiple nodes. Is it fine to use one node to recalculate the fingerprints for all variables in the system?

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj (hunk 5 above)
Comment (ygliuvt) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: Done.

File: bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj (hunk 1 above)
Comment (ygliuvt) [reply bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj]: Done.

File: bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj (hunk 9 above)
Comment (ygliuvt) [reply bootstrap-app/src/cmr/bootstrap/services/bootstrap_service.clj]: Good point. I have switched it to be dispatch through message queue on a per provider basis.

File: system-int-test/src/cmr/system_int_test/utils/bootstrap_util.clj (hunk 7 above)
Comment (ygliuvt) [reply system-int-test/src/cmr/system_int_test/utils/bootstrap_util.clj]: I didn't do it here  because I am passing in the provider id through parameters, which is easier to invoke, but not the POST standard where the payload should be in the body. I think I am going to change the api to pass in the provider id in the request body.

File: system-int-test/src/cmr/system_int_test/utils/url_helper.clj (hunk 3 above)
Comment (ygliuvt) [reply system-int-test/src/cmr/system_int_test/utils/url_helper.clj]: Done.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj (hunk 4 above)
Comment (ygliuvt) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: These are done for performance optimization. The existing code for finding concepts with params are not efficient for large volume concepts such as variables. We should consolidate the params search code with this function where the filtering of latest revision of concept is done through sql rather than code, and is much more efficient. I don't have time to complete the refactoring within this PR. We will do it in a separate issue later.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj (hunk 6 above)
Comment (ygliuvt) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/search.clj]: As we have discussed, we will work the refactoring issue (CMR-5704) later.

File: bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  [system provider]
+  (info "Updating fingerprints of variables for provider: " (:provider-id provider))
+  (let [db (helper/get-metadata-db-db system)
+        {:keys [provider-id]} provider
+        params {:concept-type :variable
+                :provider-id provider-id}
+        variable-batches (db/find-concepts-in-batches-with-stmt db
+                                                                provider
+                                                                params
+                                                                (find-variables-sql provider-id)
```
Comment (indiejames) [inline bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj]: Is `provider-id` sanitized? They way we construct the SQL using `format` seems like it could leave us open to SQL injection (I'm assuming `provider-id` is a POST parameter).

File: bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  "Defines the ending part of the string to construct find variables sql statement."
+  (str "group by concept_id) b"
+       " where a.concept_id = b.concept_id and a.revision_id = b.rid and a.deleted = 0"))
+
+(defn- find-variables-sql
+  "Returns the sql statement to find the latest variable concept revisions that are not deleted."
+  ([]
+   (str find-variables-sql-part1 find-variables-sql-part2))
+  ([provider-id]
+   (format "%s where provider_id='%s' %s"
```
Comment (indiejames) [inline bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj]: This sees dangerous. Shouldn't we parameterize `provider-id` instead?

**Comment by indiejames** [comment]
Not requesting actual changes - just clarification on my comments.

File: bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj (hunk 10 above)
Comment (ygliuvt) [reply bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj]: The `provider-id` comes from POST body and it has to be validated before it gets here. See validation code at `cmr.bootstrap.data.fingerprint` line 68 and test at `cmr.system-int-test.bootstrap.fingerprint-test` line 159.

File: bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj (hunk 11 above)
Comment (ygliuvt) [reply bootstrap-app/src/cmr/bootstrap/data/fingerprint.clj]: See above. The `provider-id` has to be a validate id that map to an existing provider in CMR before it can get here.
